StortSpelProjekt/Engine/src/Misc/AssetLoader.cpp

```diff
@@ -129,6 +135,214 @@ Model* AssetLoader::LoadModel(const std::wstring& path)
+		normals[2] = neighbourTangents[3].cross(&neighbourTangents[2]);
+		normals[3] = neighbourTangents[0].cross(&neighbourTangents[3]);
+
+		vertexNormal = normals[0] + normals[1] + normals[2] + normals[3];
+		vertexNormal.normalize();
+
+		vertices[i].normal.x = -vertexNormal.x;
+		vertices[i].normal.y = -vertexNormal.y;
+		vertices[i].normal.z = -vertexNormal.z;
+	}
+
+	// Calculate and store indices
+	std::vector<unsigned int> indices;
+	unsigned int nrOfTriangles = (tex->GetWidth() - 1) * (tex->GetHeight() - 1) * 2;
+	unsigned int nrOfIndices = nrOfTriangles * 3;
+	unsigned int toProcess = vertices.size() - tex->GetWidth();
+
+	indices.reserve(nrOfIndices);
+
+	for (unsigned int i = 0; i < toProcess; i++)
+	{
+		// calculate the indices for each triangle.
+		// they are set up in the order upper left, lower left and upper right and  for the first triangle. 
+		// for the second triangle the indices are set up as lower left, lower right and upper right.
+
+		// First triangle
+		indices.push_back(i);
+		indices.push_back(i + tex->GetWidth());
+		indices.push_back(i + 1);
+
+		// Second triangle
+		indices.push_back(i + tex->GetWidth());
+		indices.push_back(i + tex->GetWidth() + 1);
+		indices.push_back(i + 1);
+
+		// Make sure that we dont create triangles from the border. If so add one more step.
+		/*
+				Border
+				|
+			*-*-*
+			|/|/|
+			*-*-*
+			|/|/|
+			*-*-*
+		*/
+		if (((i + 2) % tex->GetWidth()) == 0)
+		{
+			i++;
+		}
+	}
+
+	// Indices finished! Move on to tangents.
+	float3 edge1 = { 0,0,0 };
+	float3 edge2 = { 0,0,0 };
+	float2 deltaUV1 = { 0,0};
+	float2 deltaUV2 = { 0,0 };
+	//Reuse neighbours array from normals!
+	neighbours[0] = { 0 };
+	neighbours[1] = { 0 };
+	neighbours[2] = { 0 };
+	float2 uv[3] = { 0 };
+	float f = 0;
+
+	for (unsigned int i = 0; i < nrOfIndices - 3; i++)
+	{
+		neighbours[0] = { vertices[indices[i]].pos.x,	  vertices[indices[i]].pos.y,	  vertices[indices[i]].pos.z };
+		neighbours[1] = { vertices[indices[i + 1]].pos.x, vertices[indices[i + 1]].pos.y, vertices[indices[i + 1]].pos.z };
+		neighbours[2] = { vertices[indices[i + 2]].pos.x, vertices[indices[i + 2]].pos.y, vertices[indices[i + 2]].pos.z };
+		
+		uv[0] = { vertices[indices[i]].uv.x,	 vertices[indices[i]].uv.y	};
+		uv[1] = { vertices[indices[i + 1]].uv.x, vertices[indices[i + 1]].uv.y };
+		uv[2] = { vertices[indices[i + 2]].uv.x, vertices[indices[i + 2]].uv.y };
+
+		edge1 = neighbours[1] - neighbours[0];
+		edge2 = neighbours[2] - neighbours[0];
+		deltaUV1 = uv[1] - uv[0];
+		deltaUV2 = uv[2] - uv[0];
+
+		f = 1.0f / (deltaUV1.x * deltaUV2.y - deltaUV2.x * deltaUV1.y);
+		for (unsigned int j = 0; j < 3; j++)
+		{
+			vertices[indices[i + j]].tangent.x = f * (deltaUV2.y * edge1.x - deltaUV1.y * edge2.x);
+			vertices[indices[i + j]].tangent.y = f * (deltaUV2.y * edge1.y - deltaUV1.y * edge2.y);
+			vertices[indices[i + j]].tangent.z = f * (deltaUV2.y * edge1.z - deltaUV1.y * edge2.z);
+		}
```

Comment:
I ran your test scene in both debug & release mode. In release all went well, but in debug the program never got past your "loadHeightMap"-functionCall in main. I paused the program and the programPointer was in this loop. Maybe something is wrong here?

Reply:
How long did you wait? For me it took just over a minute for everything to load on debug

Reply:
The test heightmap is at this moment a model with 2048*2048 = 4194304 vertices that need to be processed. Each vertex needs to have a normal and a tangent calculated. Each triangle also needs to process three indexes for vertices resulting in calculations being made for 25141254 indices.
I'm happy to take another look at it but I don't think I can optimize it that much.